docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1058 @@
+    + [Bit operations](#bit-operations)
+    + [Equality](#equality)
+    + [Comparison](#comparison)
+    + [Math](#math)
+    + [Conversion](#conversion)
+    + [Widening and Narrowing](#widening-and-narrowing)
+    + [Shuffle](#shuffle)
+  * [Summary](#summary)
+    + [Best practices](#best-practices)
+
+TL;DR: Go to [Summary](#summary)
+
+# Introduction to vectorization with Vector128 and Vector256
+
+Vectorization is an art of converting an algorithm from operating on a single value at a time to operating on a set of values (vector). It can greatly improve performance at a cost of increased code complexity.
+
+In the recent releases, .NET has introduced plenty of APIs for vectorization. Vast majority of them were hardware specific. It required the users to provide implementation per processor architecture (x64 and/or arm64), with a possibility to use the most optimal instructions for hardware that is executing the code.
+
+.NET 7 introduced a set of new APIs for `Vector128<T>` and `Vector256<T>` that aim for writing hardware-agnostic, and cross platform vectorized code. The purpose of this document is to introduce the readers to the new APIs and provide a set of best practices.
+
+## Code structure
+
+`Vector128<T>` represents a 128-bit vector of type `T`. `T` is constrained to specific primitive types:
+
+* `byte` and `sbyte` (8 bits).
+* `short` and `ushort` (16 bits).
+* `int`, `uint` and `float` (32 bits).
+* `long`, `ulong` and `double` (64 bits).
+* `nint` and `unit` (32 or 64 bits, depending on the architecture)
+
+Each `Vector128` operation allows to operate on: 16 (s)bytes, 8 (u)shorts, 4 (u)ints/floats and 2 (u)longs/double(s).
+
+```
+------------------------------128-bits---------------------------
+|             64                |               64              |
+-----------------------------------------------------------------
+|      32       |      32       |      32       |      32       |
+----------------------------------------------------------------|
+|  16   |  16   |  16   |  16   |  16   |  16   |  16   |  16   |
+-----------------------------------------------------------------
+| 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 | 8 |
+-----------------------------------------------------------------
+```
+
+`Vector256<T>` is twice as big as `Vector128<T>`, so when it is hardware accelerated, and the data is large enough we should prefer it over a `Vector128<T>`. To check the acceleration, we need to use `Vector128.IsHardwareAccelerated` and `Vector256.IsHardwareAccelerated` properties.
+
+We also must account for the size of the input. It needs to be at least of the size of a single vector to be able to execute the vectorized code path. `Vector128<T>.Count` and `Vector256<T>.Count` return the size of a vector of given type in bytes.
+Both APIs are turned into constants (no method call is required to retrieve the information) by the Just-In-Time compiler. It's not true for pre-compiled code (NativeAOT).
```

Comment:
Good point! Would following suggestion make it more clear?

```suggestion
Both APIs are turned into constants (no method call is required to retrieve the information) by the Just-In-Time compiler. In case of pre-compiled code (NativeAOT) it's not true for `IsHardwareAccelerated` property, as the required information is not available at compile time.
```